crates/pipeline-manager/src/runner/main.rs

```diff
@@ -0,0 +1,246 @@
+pub const READY_CHECK_POLL_PERIOD: Duration = Duration::from_millis(2_000);
+
+/// Number of request tries before giving up.
+pub const READY_CHECK_HTTP_RETRIES: u64 = 8;
+
+/// Type alias shorthand for the pipelines state the runner manager maintains and interacts with.
+type PipelinesState = BTreeMap<PipelineId, (Arc<Notify>, Sender<Sender<LogMessage>>)>;
+
+/// Returns whether the runner is healthy.
+/// The health check consults the continuous probe of database reachability.
+#[get("/healthz")]
+async fn get_healthz(data: web::Data<Arc<Mutex<Probe>>>) -> Result<impl Responder, ManagerError> {
+    data.lock().await.status_as_http_response()
+}
+
+/// Produces a stream of logs which it fetches from the receiver.
+/// If the channel is closed, the receiver will return and the stream will end.
+async fn logs_stream(
+    mut receiver: Receiver<LogMessage>,
+) -> impl Stream<Item = Result<web::Bytes, actix_web::Error>> {
+    stream! {
+        let mut ended_with_reason = false;
+        loop {
+            match receiver.recv().await {
+                None => {
+                    break;
+                }
+                Some(message) => {
+                    match message {
+                        LogMessage::Line(line) => {
+                            yield Ok(actix_web::web::Bytes::from(format!("{line}\n")));
+                        }
+                        LogMessage::End(reason) => {
+                            ended_with_reason = true;
+                            yield Ok(actix_web::web::Bytes::from(format!("LOGS STREAM END: {reason}\n")));
+                            break;
+                        }
+                    }
+
+                }
+            }
+        }
+        // Unexpected ending of logs stream generally indicates a panic occurred in the logging thread
+        if !ended_with_reason {
+            yield Ok(actix_web::web::Bytes::from("LOGS STREAM ENDED UNEXPECTEDLY\n"))
+        }
+    }
+}
+
+/// Retrieves as a stream the logs of a particular pipeline identified by its identifier.
+#[get("/logs/{pipeline_id}")]
+async fn get_logs(
+    data: web::Data<Arc<Mutex<PipelinesState>>>,
+    req: HttpRequest,
+) -> Result<impl Responder, ManagerError> {
+    // Parse pipeline identifier
+    let pipeline_id = parse_string_param(&req, "pipeline_id")?;
+    let pipeline_id =
+        PipelineId(
+            Uuid::from_str(&pipeline_id).map_err(|e| ManagerError::InvalidUuidParam {
+                value: pipeline_id.clone(),
+                error: e.to_string(),
+            })?,
+        );
+
+    // Attempt to follow the logs and return them in a streaming response
+    match data.lock().await.get(&pipeline_id) {
+        None => Ok(HttpResponse::NotFound().finish()),
+        Some((_, follow_request_sender)) => {
+            let (sender, receiver) = channel::<LogMessage>(CHANNEL_LOGS_BUFFER_BYTE);
+            follow_request_sender.send(sender).await.unwrap();
```

Comment:
Is there no situation where the other end will close and this unwrap will bring the whole manager down? Should we fail more gracefully here?

Reply:
I overlooked this unwrap, thanks! Will fix it with an appropriate error.

Reply:
Fixed